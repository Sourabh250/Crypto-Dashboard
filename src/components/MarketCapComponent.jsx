import React, { useEffect, useState } from "react";
import { IoTriangle } from "react-icons/io5";
import { useSelector } from "react-redux";
import numeral from "numeral";

function MarketCapComponent() {
  const [coinState, setCoinState] = useState({ coins: [], symbol: "$" });
  const [isLoading, setIsLoading] = useState(false);
  const reduxData = useSelector((state) => state.currency).toLowerCase();

  // Function to get currency symbol
  const getSymbol = (data) => {
    const Symbols = {
      usd: "$",
      inr: "₹",
      eur: "€",
      jpy: "¥",
      gbp: "£",
    };
    return Symbols[data];
  };

  // Fetch cryptocurrency data
  useEffect(() => {
    const fetchData = async () => {
      setIsLoading(true);
      const apiKey = "CG-7TzT57ZLhugEtsmWZnyA8d1H";
      try {
        const url = `https://api.coingecko.com/api/v3/coins/markets?vs_currency=${reduxData}&order=market_cap_desc&page=1&price_change_percentage=24h&x_cg_demo_api_key=${apiKey}`;
        const response = await fetch(url);
        if (!response.ok) {
          throw new Error("Failed to fetch. Please try again later.");
        }
        const data = await response.json();
        setCoinState({ coins: data, symbol: getSymbol(reduxData) });
        setIsLoading(false);
      } catch (error) {
        console.error(error);
      }
    };

    fetchData();
  }, [reduxData]);

  // Function to format market cap
  const formatMcap = (value) => {
    return numeral(value).format("0.0a").toUpperCase();
  };

  // Render the component
  return (
    <div className="bg-gray-100 h-full w-full md:w-1/4 rounded-lg shadow-lg py-3 lg:py-6">
      <h1 className="text-lg lg:text-xl font-semibold px-4">
        Cryptocurrency by
        <br />
        market cap
      </h1>

        {/* Loading spinner or list of cryptocurrencies */}
        {isLoading ? (
          <div className="flex justify-center items-center h-full">
          <div className="spinner"></div>
        </div>
        ) : (
          <div className="flex flex-col text-sm">
            {coinState.coins.slice(0, 7).map((data, index) => (
              <div
                className="flex justify-between items-center px-4 py-3 lg:py-6 border-b"
                key={index}
              >
                <div>
                  <h1 className="font-medium">{data.name}</h1>
                  <h2 className="font-extralight">
                    M.Cap {coinState.symbol}
                    {formatMcap(data.market_cap)}
                  </h2>
                </div>
                <div className="flex justify-center items-center gap-1">

                  {/* Display price change percentage with arrow */}
                  {data.price_change_percentage_24h > 0 ? (
                    <IoTriangle color="green" />
                  ) : (
                    <IoTriangle
                      color="red"
                      style={{ transform: "rotate(180deg)" }}
                    />
                  )}
                  {Math.abs(data.price_change_percentage_24h).toFixed(2)}%
                </div>
              </div>
            ))}
          </div>
        )}
    </div>
  );
}

export default MarketCapComponent;
