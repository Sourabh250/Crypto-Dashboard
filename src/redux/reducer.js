function reducer(state = {currency : "USD", selectedCurrencies: ["Ethereum"]},action) {
    switch(action.type) {
        // Update the selected currency in the state
        case "SET_CURRENCY":
            return {
                ...state,
                currency: action.payload
            };
        // Add a new cryptocurrency to the list of selected cryptocurrencies
        case "ADD_CRYPTO_CURRENCY" :
            return {
                ...state,
                selectedCurrencies: [...state.selectedCurrencies, action.payload],
            };
        // Remove a cryptocurrency from the list of selected cryptocurrencies
        case "REMOVE_CRYPTO_CURRENCY" :
            return {
                ...state,
                selectedCurrencies: state.selectedCurrencies.filter(data => data !== action.payload),
            };
        // Modify the list of selected cryptocurrencies to only contain the specified cryptocurrency
        case "MODIFY_CRYPTO_CURRENCY" :
            return {
                ...state,
                selectedCurrencies: [action.payload],
            }
        // Return the current state if the action type is not recognized
        default:
            return state;
    }
}

export default reducer;