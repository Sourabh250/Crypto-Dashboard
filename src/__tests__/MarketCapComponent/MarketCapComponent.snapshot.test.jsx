import React from "react";
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from '../../redux/reducer';
import MarketCapComponent from '../../components/MarketCapComponent';

const store = createStore(reducer);

test('MarketCapComponent snapshot', () => {
    const tree = renderer.create(
        <Provider store={store}>
            <MarketCapComponent />
        </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
})