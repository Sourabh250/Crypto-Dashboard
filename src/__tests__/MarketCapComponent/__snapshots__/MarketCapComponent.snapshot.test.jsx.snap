// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketCapComponent snapshot 1`] = `
<div
  className="bg-gray-100 h-full w-full md:w-1/4 rounded-lg shadow-lg py-3 lg:py-6"
>
  <h1
    className="text-lg lg:text-xl font-semibold px-4"
  >
    Cryptocurrency by
    <br />
    market cap
  </h1>
  <div
    className="flex flex-col text-sm"
  />
</div>
`;
